### DISCUSSION TITLE
ComputeIsotropicElasticityTensor from element-based imported youngs_modulus and poissons_ratio

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#20188] [Q&A Modules: Solid mechanics] [answered]
Hi guys,

I'd like to ComputeIsotropicElasticityTensor from element-based imported youngs_modulus and poissons_ratio 

I can import heterogenous youngs_modulus and poissons_ratio using `ElementPropertyReadFile` (to read .csv file) and `GenericConstantArray` (this is successfully done without a problem).
but in `ComputeIsotropicElasticityTensor ` it only takes a single value of  youngs_modulus and poissons_ratio such that

```
  params.addParam<Real>("poissons_ratio", -1, "Poisson's ratio for the material.");
  params.addParam<Real>("youngs_modulus", -1, "Young's modulus of the material.");
.....
  if (_youngs_modulus_set && _poissons_ratio_set)
  {
    _Cijkl.fillSymmetricIsotropicEandNu(_youngs_modulus, _poissons_ratio);
    _effective_stiffness_local =
        std::max(std::sqrt((_youngs_modulus * (1 - _poissons_ratio)) /
                           ((1 + _poissons_ratio) * (1 - 2 * _poissons_ratio))),
                 std::sqrt(_youngs_modulus / (2 * (1 + _poissons_ratio))));
    return;
  }
.....
  // Assign elasticity tensor at a given quad point
  _elasticity_tensor[_qp] = _Cijkl;

  // Assign effective stiffness at a given quad point
  _effective_stiffness[_qp] = _effective_stiffness_local;
```

is there a way to work around this? my imported youngs_modulus and poissons_ratio would be in this format:
`_youngs_modulus2(getMaterialProperty<Real>("_youngs_modulus2")`  and `_poissons_ratio2(getMaterialProperty<Real>("_poissons_ratio2")` >>> `_youngs_modulus2[_qp]` and `_poissons_ratio2[_qp]` which I cannot simply just replace to the above

to summarise, the question would be, how do I compute  _elasticity_tensor[_qp] and _effective_stiffness[_qp] based on element-based imported youngs_modulus and poissons_ratio.
..................................................................................
you might say why don't I just use a block-restricted approach to do this, it won't work if I want to control mining activities such as excavation and backfill on a highly heterogeneous model, modifying the mesh is also not ideal in our workflow.

![image](https://user-images.githubusercontent.com/79562439/151906113-ca4ed853-1e50-4d0e-ae23-5da9048197d5.png)

Since MOOSE doesn't allow overlapping blocks, I plan to use a block-restricted approach to control mining activities (white blocks) instead by multiplying the imported properties by some 'prefactor'.

Thank you very much

Kind regards,
Traiwit

**Comment by GiudGiud** [comment]
Hello

Unless you can find a pre-exisiting one, 
You'll probably have to make your own class which allows a material property for these coefficients, instead of the constant values which this one allows. 
It should be fairly easy though

Guillaume

**Reply by Traiwit** [reply]
Hi @GiudGiud,

so the original way of assigning the constant value is this 

```
  _Cijkl.fillSymmetricIsotropicEandNu(_youngs_modulus, _poissons_ratio);
    _effective_stiffness_local =
        std::max(std::sqrt((_youngs_modulus * (1 - _poissons_ratio)) /
                           ((1 + _poissons_ratio) * (1 - 2 * _poissons_ratio))),
                 std::sqrt(_youngs_modulus / (2 * (1 + _poissons_ratio))));
    return;

```
not sure what does `fillSymmetricIsotropicEandNu` do, but when I modified it to:

```
    _Cijkl.fillSymmetricIsotropicEandNu(_youngs_modulus_csv[_qp], _poissons_ratio_csv[_qp]);
    _effective_stiffness_local =
        std::max(std::sqrt((_youngs_modulus_csv[_qp] * (1 - _poissons_ratio_csv[_qp])) /
                           ((1 + _poissons_ratio_csv[_qp]) * (1 - 2 * _poissons_ratio_csv[_qp]))),
                 std::sqrt(_youngs_modulus_csv[_qp] / (2 * (1 + _poissons_ratio_csv[_qp]))));
    return;
```

i got this error:

```
> moosebe-opt: malloc.c:4036: _int_malloc: Assertion `(unsigned long) (size) >= (unsigned long) (nb)' failed.
> corrupted size vs. prev_size
> 
> ===================================================================================
> =   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
> =   PID 799677 RUNNING AT moose-System-Product-Name
> =   EXIT CODE: 134
> =   CLEANING UP REMAINING PROCESSES
> =   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
> ===================================================================================
> YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Aborted (signal 6)
> This typically refers to a problem with your application.
> Please see the FAQ page for debugging suggestions
```

**Reply by GiudGiud** [reply]
Hi

The location where you modified this is a setup function. It s not being ran on every quadrature point, and the reason why is that this class defines constant isotropic material properties, so you only need to compute them once. _qp is not defined at this point, and the material properties have not been initialized. 

If you want to be recomputing things on every quadrature point, with a _qp index into a material property, then you need to be recomputing things in `computeQpElasticityTensor()` in that material. 

But it seems @josebastiase has already worked on this, so please have a look at his material. 

Guillaume

**Comment by Traiwit** [comment]
that might not be a problem at all, the problem could come from the way I define those array of Material properties:

in .h
```
const MaterialProperty<Real> & _poissons_ratio_csv;
const MaterialProperty<Real> & _youngs_modulus_csv;
```
in .C
```
_poissons_ratio_csv(this->template getMaterialProperty<Real>("poissons_ratio_csv")),
_youngs_modulus_csv(this->template getMaterialProperty<Real>("youngs_modulus_csv")),
```

I've never modified any file with AD option, i think i probs stuff it up at the initial step

**Reply by GiudGiud** [reply]
this looks fine, for non AD work

**Comment by josebastiase** [comment, marked as answer]
This may be useful (https://github.com/josebastiase/RHEA) just copy and paste IsotropicElasticModulusFromVar.C and IsotropicElasticModulusFromVar.h in your tensor mechanics materials. Depending on what type of physics are you using in PorousFlow, you may also want the Biot modulus.

**Reply by Traiwit** [reply]
Hi @josebastiase,

thank you, i think i almost get it, here is my modified version:

```
#include "ComputeIsotropicElasticityTensorBE.h"

registerMooseObject("moosebeApp", ComputeIsotropicElasticityTensorBE);

template <>
InputParameters
validParams<ComputeIsotropicElasticityTensorBE>()
{
  InputParameters params = validParams<ComputeElasticityTensorBase>();
  params.addClassDescription("Compute a isotropic, heterogeneous elasticity tensor using the "
			     "spatially-varying bulk_modulus and shear_modulus values.");
params.addParam<MaterialPropertyName>("youngs_modulus_csv",-1,"The coupled variable which provides the force");
params.addParam<MaterialPropertyName>("poissons_ratio_csv",-1,"The coupled variable which provides the force");
  return params;
}

ComputeIsotropicElasticityTensorBE::ComputeIsotropicElasticityTensorBE(const InputParameters & parameters)
  : ComputeElasticityTensorBase(parameters),
  _poissons_ratio_csv(getMaterialProperty<Real>("poissons_ratio_csv")),
  _youngs_modulus_csv(getMaterialProperty<Real>("youngs_modulus_csv"))
{
   issueGuarantee(_elasticity_tensor_name, Guarantee::ISOTROPIC);
}

void
ComputeIsotropicElasticityTensorBE::computeQpElasticityTensor()
{
 // Fill elasticity tensor

  _Cijkl.fillSymmetricIsotropicEandNu(_poissons_ratio_csv[_qp], _youngs_modulus_csv[_qp]);
  _elasticity_tensor[_qp] = _Cijkl;
}
```

and this is the error I got after 1 iteration of nl

```
>  58 Linear |R| = 9.723134e-07
>      59 Linear |R| = 9.491814e-07
>      60 Linear |R| = 9.263357e-07
>      61 Linear |R| = 9.153495e-07
>      62 Linear |R| = 9.046562e-07
>      63 Linear |R| = 8.917094e-07
>      64 Linear |R| = 8.703339e-07
> Cannot take square root of a number less than or equal to zero in the calculation of C3_test for the Rashid approximation for the rotation tensor. This zero or negative number may occur when elements become heavily distorted.
> 
> A MooseException was raised during Auxiliary variable computation.
> The next solve will fail, the timestep will be reduced, and we will try again.
> 
> Cannot take square root of a number less than or equal to zero in the calculation of C1 for the Rashid approximation for the rotation tensor. This zero or negative number may occur when elements become heavily distorted.
> 
> A MooseException was raised during Auxiliary variable computation.
> The next solve will fail, the timestep will be reduced, and we will try again.
> 
> Cannot take square root of a number less than or equal to zero in the calculation of C1 for the Rashid approximation for the rotation tensor. This zero or negative number may occur when elements become heavily distorted.
> 
> A MooseException was raised during Auxiliary variable computation.
> The next solve will fail, the timestep will be reduced, and we will try again.

Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
Solve Did NOT Converge!
```

I'm not super certain how `fillSymmetricIsotropicEandNu` works, but it does seem like it doesn't like the way I input `_poissons_ratio_csv[_qp], _youngs_modulus_csv[_qp]`

Thanks!
Traiwit

@GiudGiud  @WilkAndy I tag you two as well, just in case you are familiar with this issue

**Reply by Traiwit** [reply]
all good now, so I just change the last bit to this: 

```
void
ComputeIsotropicElasticityTensorBE::computeQpElasticityTensor()
{
 // Fill elasticity tensor

 const Real E = _youngs_modulus_csv[_qp];
 const Real nu = _poissons_ratio_csv[_qp];



  _Cijkl.fillSymmetricIsotropicEandNu(E, nu);
  _elasticity_tensor[_qp] = _Cijkl;
}
```

Thank you everyone!

**Reply by manikumd22** [reply]
Hi @josebastiase/ @Traiwit  

I need little help from your side, anyone please have a look at it. 

I have gone through whole discussion and make some progress for reading csv file for material property. However, I'm facing some issues while defining it in input file, this is the way I am reading the value of lambda constant from csv file, but I am getting an error
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'lambda1', requested by 'compute_stress' is not defined on block surface
Material property 'lambda1', requested by 'compute_stress_face' is not defined on block surface
Material property 'lambda1', requested by 'compute_stress_neighbor' is not defined on block surface

E_values.csv:
5, 10000.0


 [Functions]
    [./lambda1]
      type = PiecewiseConstantFromCSV
      read_prop_user_object = 'reader_block'
      read_type = 'block'
      column_number = '2'
    [../]
  []

  [UserObjects]
    [reader_block]
      type = PropertyReadFile
      prop_file_name = 'E_values.csv'
      read_type = 'block'
      nprop = 1
      nblock = 5
    []
  []


  [Materials]
    [compute_stress]
      type = ComputeNeoHookeanStress
      lambda = lambda1
      mu = 8300.33333888888926
    []
    [compute_strain]
      type = ComputeLagrangianStrain
    []
  []

**Reply by GiudGiud** [reply]
answer is on the other thread. 
Please dont double post.

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I am sorry, I didn't realize that
